### ISSUE TITLE
Codestyle: disallow same variable name with different casing

### ISSUE THREAD

**Issue by kendrahavens** [dotnet/roslyn#38531]
[Original devcomm ticket](https://developercommunity.visualstudio.com/content/idea/697873/have-a-setting-to-allowdisallow-veriables-with-sam.html)

> Have a setting to allow/disallow variables with same name but different letter cases

*Example*
> I accidentally had two namespaces, GpsData and GPSData, and it allowed them both and I couldn't find the classes of the one in the other, thinking there was only one namespace. Same with variables.

**Comment by markus27183** [comment]
This is a great idea.

**Comment by CyrusNajmabadi** [comment]
This can be done by users writing their own analyzer.

**Comment by markus27183** [comment]
Yes I can also write my own operating system, or my own IDE or runtime
environment.  But if you want a customer like me to like your product more
you will add this feature.

**Comment by CyrusNajmabadi** [comment]
@markus27183 i'd be happy to walk you through creating such an analyzer.  With that info you could then add additional analyzers in the future if you run into more stuff you want to check for (or you can change this particular analyzer if you run into cases you don't want it to trigger on).

LMK!

**Comment by sharwell** [comment]
Design review conclusion: We aren't sure what the "right" approach here is, so we're interested in feedback from users regarding exactly how this analysis would be expected to behave. Some items to consider:

1. CLS compliance does already check that _exposed_ names do not differ only by case

2. An FXCop analyzer existed for this, but was not ported to FXCop Analyzers in dotnet/roslyn-analyzers

3. We need to be careful about code styles that intentionally use names that differ only by case, such as the following naming for properties with backing fields:

    ```csharp
    private int propertyValue;
    public int PropertyValue {
      get => propertyValue;
      set => propertyValue = value;
    }
    ```

4. Some possibilities for an analyzer are:
    * Only check namespaces, or perhaps namespaces and types
    * Only check names defined in the same scope
    * Check names defined in the same scope, but intentionally omit warnings for the field/property naming seen above

Other options certainly exist.

**Comment by CyrusNajmabadi** [comment]
Not planning on doing this.  This is exactly what the analyzer system is for.  So taht people can add rules like this.
